Combine limit and offset to `fetch` and `skip` and implement physical plan support

ming535 opened the pull request:
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2551, also closes this pr https://github.com/apache/arrow-datafusion/pull/2629

 # Rationale for this change
1. Follow up the discussion here https://github.com/apache/arrow-datafusion/pull/2638, this pr combines "limit" and "offset" together. This PR changes the api to
`limit(skip: Option<usize>, fetch: Option<usize>)` where `skip` is the number of rows to skip, `fetch` is the number of
rows to return.

When designing the api, I looked into two alternatives:
a) calcite: https://calcite.apache.org/javadocAggregate/org/apache/calcite/tools/RelBuilder.html#limit(int,int)
b) substrait: https://substrait.io/relations/logical_relations/#fetch-operation

substrait's `fetch` seems nice,  but the change from `limit` to `fetch` seems too much in our current code base, so I adopted
the api in calcite.


2. implement the logic of "offset" in physical plan.

# What changes are included in this PR?
1. api refactoring: combine "limit" and "offset" together. 
4. implement the logic of "offset" in physical plan.

# Are there any user-facing changes?
yes.

Ted-Jiang commented on datafusion/core/src/physical_plan/limit.rs:
```diff
@@ -176,17 +184,18 @@ impl ExecutionPlan for GlobalLimitExec {
             Statistics {
                 num_rows: Some(nr), ..
-            } if nr <= self.limit => input_stats,
+            } if nr - skip <= self.fetch.unwrap_or(usize::MAX) => input_stats,
```
nit: like
```
if the input is greater than the limit, the num_row will be the limit
```
and 
```
// if the input is greater than the limit, the num_row will be greater
// than the limit because the partitions will be limited separatly
// the statistic
```
limit should change to fetch

ming535 replied:
refactored the logic.

Ted-Jiang reviewed (commented):
I skimmed this PR, it looks really great! 👍  i will read carefully this evening.

alamb commented:
I am reviewing this carefully now

alamb commented on datafusion/optimizer/src/limit_push_down.rs:
```diff
@@ -468,16 +477,14 @@ mod test {
-        \n  Offset: 10\
-        \n    Projection: #test.a\
-        \n      TableScan: test projection=None, limit=1010";
+        let expected = "Limit: skip=10, fetch=1000\
```
I reviewed the tests carefully in this file and they look good to me 👍

alamb commented on datafusion/optimizer/src/limit_push_down.rs:
```diff
@@ -490,14 +497,14 @@ mod test {
 
-        let expected = "Offset: 10\
-        \n  Limit: 1000\
+        let expected = "Limit: skip=10, fetch=None\
```
Perhaps a future optimization could consolidate these together into a single `Limit` -- I don't think it is particularly important, just to be clear

alamb reviewed (commented):
Thank you for the contribution @ming535  -- this is looking really good, but I think there is a problem with multi partition plans (I left a reproducer inline)

The overall the structure of the the code and the description of this PR are 👌  very nice. It was a pleasure to read.

I like the choice of [calcite](https://calcite.apache.org/javadocAggregate/org/apache/calcite/tools/RelBuilder.html#limit(int,int)) 's API

ming535 commented on datafusion/core/src/physical_plan/planner.rs:
```diff
@@ -906,15 +905,14 @@ impl DefaultPhysicalPlanner {
                         // a CoalescePartitionsExec between the GlobalLimitExec and LocalLimitExec
-                        Arc::new(LocalLimitExec::new(input, limit))
+                        if let Some(fetch) = fetch {
+                            Arc::new(LocalLimitExec::new(input, *fetch))
```
fixed, and added a new test case with multi-partitioned table.

alamb replied:
Looks good now:

```
❯ select * from test limit 2 offset 1;
+----------+----------+
| column_1 | column_2 |
+----------+----------+
| 3        | 4        |
| a        | b        |
+----------+----------+
2 rows in set. Query took 0.039 seconds.
❯ 
```

ming535 commented on datafusion/core/src/physical_optimizer/repartition.rs:
```diff
@@ -449,13 +450,13 @@ mod tests {
+            "LocalLimitExec: fetch=100",
             // Expect no repartition to happen for local limit
             "ParquetExec: limit=None, partitions=[x], projection=[c1]",
         ];
```
done

ming535 commented on datafusion/sql/src/planner.rs:
```diff
@@ -1212,57 +1206,59 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+            return Ok(input);
+        }
+
+        let skip = match skip {
```
tried this, `map` is complaining about the type. When error happens in parsing the `i64` value, it cannot figure out the right return type.
<img width="1231" alt="Screen Shot 2022-06-08 at 20 24 39" src="https://user-images.githubusercontent.com/7576663/172615511-06647bc5-9ccb-44d7-a119-0e2fde02556a.png">

alamb replied:
This compiled for me:

```rust

        let skip = skip
            .map(|skip_expr| {
                let skip = match self.sql_to_rex(
                    skip_expr.value,
                    input.schema(),
                    &mut HashMap::new(),
                )? {
                    Expr::Literal(ScalarValue::Int64(Some(s))) => {
                        if s < 0 {
                            return Err(DataFusionError::Plan(format!(
                                "Offset must be >= 0, '{}' was provided.",
                                s
                            )));
                        }
                        Ok(s as usize)
                    }
                    _ => Err(DataFusionError::Plan(
                        "Unexpected expression in OFFSET clause".to_string(),
                    )),
                }?;
                Ok(skip)
            })
            .transpose()?;

```

But I don't think it is all that much more readable than what you have in this PR

ming535 commented on datafusion/optimizer/src/limit_push_down.rs:
```diff
@@ -43,61 +41,73 @@ impl LimitPushDown {
+                } => {
+                    if let Some(fetch) = current_fetch {
+                        // extend ancestor's fetch
+                        let ancestor_fetch =
```
I think it is. I added some more comments in the code.

ming535 commented on datafusion/optimizer/src/limit_push_down.rs:
```diff
@@ -43,61 +41,73 @@ impl LimitPushDown {
-                    new_ancestor_offset,
-                    new_ancestor_limit,
+                    Ancestor::FromLimit {
+                        skip: *current_skip,
```
No, change `current_skip` will violate the semantic in the tree.  Current node's `skip` is never updated.

ming535 commented:
I just did the following changes:
1. fixed one bug related to partitioned table
2. fixed one bug related to statistics
3. improve code comments.

You can review it again. @alamb @Ted-Jiang

alamb commented on datafusion/core/src/physical_optimizer/repartition.rs:
```diff
@@ -408,6 +416,23 @@ mod tests {
+
+        let expected = &[
+            "GlobalLimitExec: skip=5, fetch=100",
+            "LocalLimitExec: fetch=100",
```
Don't we have to fetch at least `105` from each local partition (`skip` + `fetch`) to be sure that a sufficient number of rows is produced?

Thus I would expect this to be

LocalLimitExec: fetch=105

However, when I tried an actual query it seems to do the right thing so 👍 

```
❯ explain select * from test limit 1 offset 5;
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                        |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Limit: skip=5, fetch=1                                                                                                                      |
|               |   Projection: #test.column_1, #test.column_2                                                                                                |
|               |     TableScan: test projection=Some([column_1, column_2]), fetch=6                                                                          |
| physical_plan | GlobalLimitExec: skip=5, fetch=1                                                                                                            |
|               |   CoalescePartitionsExec                                                                                                                    |
|               |     LocalLimitExec: fetch=6                                                                                                                 |
|               |       ProjectionExec: expr=[column_1@0 as column_1, column_2@1 as column_2]                                                                 |
|               |         RepartitionExec: partitioning=RoundRobinBatch(16)                                                                                   |
|               |           CsvExec: files=[/private/tmp/csv/1.csv, /private/tmp/csv/2.csv], has_header=false, limit=Some(6), projection=[column_1, column_2] |
|               |                                                                                                                                             |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------+
```

alamb commented on datafusion/core/src/physical_plan/empty.rs:
```diff
@@ -226,4 +235,23 @@ mod tests {
     }
+
+    #[tokio::test]
+    async fn produce_one_row_multiple_partition() -> Result<()> {
```
👍

alamb commented on datafusion/core/src/physical_plan/limit.rs:
```diff
@@ -186,20 +181,31 @@ impl ExecutionPlan for GlobalLimitExec {
-                ..Default::default()
-            },
-            // if we don't know the input size, we can't predict the limit's behaviour
+            } => {
```
thank you I think this is easier to read now

alamb reviewed (approved):
I reviewed the changes in https://github.com/apache/arrow-datafusion/pull/2694/commits/f7f5318393747c481ca2ccb7968a34bc03bc7e8b and it  looks great. 

Thank you so much @ming535  🏅  -- this is really nice work

alamb commented:
I think this PR is ready to merge -- @Ted-Jiang  any other comments? Unless I hear otherwise I'll plan to merge it tomorrow.

Ted-Jiang reviewed (approved):
LGTM 👍 @ming535 Nice work!

alamb commented:
Thanks @ming535  and @Ted-Jiang  -- this is great.
